src/graph_ir.rs

```diff
@@ -1,35 +1,38 @@
 //! The internal [GraphIR] datastructure used by the compiler passes.
 //!
 use crate::{
-    buffer_allocator2::{BufferAllocator, BufferRef},
+    buffer_allocator::{BufferAllocator, BufferRef},
+    graph_helper::NodeEdges,
     input_ir::*,
     output_ir::*,
 };
-use serde::{Deserialize, Serialize};
-use std::collections::{HashMap, HashSet};
+use fnv::{FnvHashMap, FnvHashSet};
 use std::rc::Rc;
 
+#[cfg(feature = "serialize")]
+use serde::{Deserialize, Serialize};
+
 /// Internal IR used by the compiler algorithm. Built incrementally
 /// via the compiler passes.
-#[derive(Clone, Debug, Serialize, Deserialize)]
-pub struct GraphIR {
+#[cfg_attr(feature = "serialize", derive(Serialize, Deserialize))]
+#[derive(Clone, Debug)]
+pub struct GraphIR<'a> {
     /// The number of port types used by the graph.
-    pub num_port_types: usize,
+    num_port_types: usize,
     /// A table of nodes in the graph.
-    pub nodes: HashMap<u64, Node>,
-    /// A list of edges in the graph.
-    pub edges: HashMap<u64, Edge>,
+    nodes: &'a FnvHashMap<NodeID, Node>,
```

Comment:
The `GraphIR` should own its state, including these tables. I'm going to move ownership from the `graph_helper` down to this object and add a `reset` method to clear out the state without reallocating.

Reply:
Hmm. A large point of the the `AudioGraphHelper` is so it can persist the list of nodes and adjacent edges without having to re-construct that map every time you want to compile (and every time you want to check for cycles).

This optimization is especially important for cycle detection in my helper struct, since it basically runs the tarjan function for every edge you add.

Reply:
Persistence is good, but I don't want callers to need to worry about the graph_helper API (which looks special purpose, relative to meadowlark vs the simpler API which just takes nodes, edges, and port data and compiles it). 

The cycle detection can move into the GraphIR and it can handle its own persistence there.